In [1]:
import numpy as np
import math
import warnings as w
w.filterwarnings("ignore",category=RuntimeWarning)

In [2]:
w1x1 = [-5.01, -5.43, 1.08, 0.86, -2.67, 4.94, -2.51, -2.25, 5.56, 1.03]
w1x2 = [-8.12, -3.48, -5.52, -3.78, 0.63, 3.29, 2.09, -2.13, 2.86, -3.33]
w1x3 = [-3.68, -3.54, 1.66, -4.11, 7.39, 2.08, -2.59, -6.94, -2.26, 4.33]

In [3]:
w2x1 = [-0.91, 1.30, -7.75, -5.47, 6.14, 3.60, 5.37, 7.18, -7.39, -7.50]
w2x2 = [-0.18, -2.06, -4.54, 0.50, 5.72, 1.26, -4.63, 1.46, 1.17, -6.32]
w2x3 = [-0.05, -3.53, -0.95, 3.92, -4.85, 4.36, -3.65, -6.66, 6.30, -0.31]

In [4]:
w3x1 = [5.35, 5.12, -1.34, 4.48, 7.11, 7.17, 5.75, 0.77, 0.90, 3.52]
w3x2 = [2.26, 3.22, -5.31, 3.42, 2.39, 4.33, 3.97, 0.27, -0.43, -0.36]
w3x3 = [8.13, -2.66, -9.87, 5.19, 9.21, -0.98, 6.65, 2.41, -8.71, 6.43]

In [5]:
w1x1_m = np.mean(w1x1)
w1x2_m = np.mean(w1x2)
w1x3_m = np.mean(w1x3)
w2x1_m = np.mean(w2x1)
w2x2_m = np.mean(w2x2)
w2x3_m = np.mean(w2x3)
w3x1_m = np.mean(w3x1)
w3x2_m = np.mean(w3x2)
w3x3_m = np.mean(w3x3)
w1_m = np.array([w1x1_m, w1x2_m, w1x3_m])
w2_m = np.array([w2x1_m, w2x2_m, w2x3_m])
w3_m = np.array([w3x1_m, w3x2_m, w3x3_m])
w1_m_2d = np.array([w1x1_m, w1x2_m])
w2_m_2d = np.array([w2x1_m, w2x2_m])
w3_m_2d = np.array([w3x1_m, w3x2_m])
w1_m

array([-0.44 , -1.749, -0.766])

# For one Feature

In [6]:
cov_w1x1 = np.cov(w1x1)
det_w1x1 = np.array(cov_w1x1)
inverse_w1x1 = np.array([1/cov_w1x1])

cov_w2x1 = np.cov(w2x1)
det_w2x1 = np.array(cov_w2x1)
inverse_w2x1 = np.array([1/cov_w2x1])

cov_w3x1 = np.cov(w3x1)
det_w3x1 = np.array(cov_w3x1)
inverse_w3x1 = np.array([1/cov_w3x1])


In [7]:
def xmean(x , mean) :
    return (x - mean)

In [8]:
def q2a_g1(x) :
    a1 = [xmean(x, w1x1_m)]
    a1 = np.expand_dims(a1,axis=0)
    a2 = np.transpose(a1)
    a3 = (-1/2) * 1 * np.log(2 * math.pi)
    a4 = (-1/2) * np.log(det_w1x1)
    a5 = np.matmul(a1,inverse_w1x1)
    a6 = np.matmul(a5,a2)
    g = (-1/2) * a6 + a3 + a4 + np.log(1/2) #Assuming prior probability is 1/2
    return g[0]

In [9]:
def q2a_g2(x) :
    a1 = [xmean(x, w2x1_m)]
    a1 = np.expand_dims(a1,axis=0)
    a2 = np.transpose(a1)
    a3 = (-1/2) * 1 * np.log(2 * math.pi)
    a4 = (-1/2) * np.log(det_w2x1)
    a5 = np.matmul(a1,inverse_w2x1)
    a6 = np.matmul(a5,a2)
    g = (-1/2) * a6 + a3 + a4 + np.log(1/2) #Assuming prior probability is 1/2
    return g[0]

In [10]:
def q2a_g3(x) :
    a1 = [xmean(x, w3x1_m)]
    a1 = np.expand_dims(a1,axis=0)
    a2 = np.transpose(a1)
    a3 = (-1/2) * 1 * np.log(2 * math.pi)
    a4 = (-1/2) * np.log(det_w3x1)
    a5 = np.matmul(a1,inverse_w3x1)
    a6 = np.matmul(a5,a2)
    g = (-1/2) * a6 + a3 + a4 + np.log(0) #Assuming prior probability is 0
    return g[0]

In [29]:
x = np.array(-5.01)
print(np.squeeze(q2a_g1(x)))
print(np.squeeze(q2a_g2(x)))
print(np.squeeze(q2a_g3(x)))

-3.6711755718438246
-3.6861324405954443
-inf


In [12]:
def missclassified_qb():
    count1 = 0
    for a in w1x1:
        if q2a_g1(a) < max(q2a_g2(a),q2a_g3(a)):
            count1 += 1
        
    print("Missclassified Percentage in W1 is: " + str(count1*100/len(w1x1)))
    count2 = 0
    for a in w2x1:
        if q2a_g2(a) < max(q2a_g1(a),q2a_g3(a)):
            count2 += 1
    print("Missclassified Percentage in W2 is: " + str(count2*100/len(w1x2)))
    count3 = 0
    for a in w3x1:
        if q2a_g3(a) < max(q2a_g1(a),q2a_g2(a)):
            count3 += 1
    print("Missclassified Percentage in W3 is: " + str(count3*100/len(w1x3)))
    print("Missclassified Percentage in the dataset: " + str((count1+count2+count3)*100/(len(w1x1)+len(w1x2)+len(w1x3))))
missclassified_qb()

Missclassified Percentage in W1 is: 30.0
Missclassified Percentage in W2 is: 30.0
Missclassified Percentage in W3 is: 100.0
Missclassified Percentage in the dataset: 53.333333333333336


# For 2 Features

In [31]:
stacked_w1_2d = np.stack((w1x1, w1x2), axis = 0 )
cov_w1_2d = np.cov(stacked_w1_2d)
det_w1_2d = np.linalg.det(cov_w1_2d)
inverse_w1_2d = np.linalg.inv(cov_w1_2d)

stacked_w2_2d = np.stack((w2x1, w2x2), axis = 0 )
cov_w2_2d = np.cov(stacked_w2_2d)
det_w2_2d = np.linalg.det(cov_w2_2d)
inverse_w2_2d = np.linalg.inv(cov_w2_2d)


stacked_w3_2d = np.stack((w3x1, w3x2), axis = 0 )
cov_w3_2d = np.cov(stacked_w3_2d)
det_w3_2d = np.linalg.det(cov_w3_2d)
inverse_w3_2d = np.linalg.inv(cov_w3_2d)

In [14]:
def q2_g1(x, mean, inverse, det) :
    a1 = xmean(x, mean)
    a1 = np.expand_dims(a1,axis=0)
    a2 = np.transpose(a1)
    a3 = (-1/2) * x.shape[0] * np.log(2 * math.pi)
    a4 = (-1/2) * np.log(det)
    a5 = np.matmul(a1,inverse)
    a6 = np.matmul(a5,a2)
    g = (-1/2) * a6 + a3 + a4 + np.log(1/2) #Assuming prior probability is 1/2
    return g

In [15]:
def q2_g2(x, mean, inverse, det) :
    a1 = xmean(x, mean)
    a1 = np.expand_dims(a1,axis=0)
    a2 = np.transpose(a1)
    a3 = (-1/2) * x.shape[0] * np.log(2 * math.pi)
    a4 = (-1/2) * np.log(det)
    a5 = np.matmul(a1,inverse)
    a6 = np.matmul(a5,a2)
    g = (-1/2) * a6 + a3 + a4 + np.log(1/2) #Assuming prior probability is 1/2
    return g

In [16]:
def q2_g3(x, mean, inverse, det) :
    a1 = xmean(x, mean)
    a1 = np.expand_dims(a1,axis=0)
    a2 = np.transpose(a1)
    a3 = (-1/2) * x.shape[0] * np.log(2 * math.pi)
    a4 = (-1/2) * np.log(det)
    a5 = np.matmul(a1,inverse)
    a6 = np.matmul(a5,a2)
    g = (-1/2) * a6 + a3 + a4 + np.log(0) #Assuming prior probability is 0
    return g

In [17]:
# For 2 features

x = np.array([-5.01, -8.12])
print(np.squeeze(q2_g1(x, w1_m_2d, inverse_w1_2d, det_w1_2d)))
print(np.squeeze(q2_g2(x, w2_m_2d, inverse_w2_2d, det_w2_2d)))
print(np.squeeze(q2_g3(x, w3_m_2d, inverse_w3_2d, det_w3_2d)))

-6.499467023888501
-7.58483495042734
-inf


In [18]:
def missclassified_qc():
    count1 = 0
    for i in range(len(w1x1)):
        a = np.array([w1x1[i], w1x2[i]])
        if q2_g1(a, w1_m_2d, inverse_w1_2d, det_w1_2d) < max(q2_g2(a, w2_m_2d, inverse_w2_2d, det_w2_2d),q2_g3(a, w3_m_2d, inverse_w3_2d, det_w3_2d)):
            count1 += 1
    print("Missclassified Percentage in W1 is: " + str(count1*100/len(w1x1)))
    count2 = 0
    for i in range(len(w1x1)):
        a = np.array([w2x1[i], w2x2[i]])
        if q2_g2(a, w2_m_2d, inverse_w2_2d, det_w2_2d) < max(q2_g1(a, w1_m_2d, inverse_w1_2d, det_w1_2d),q2_g3(a, w3_m_2d, inverse_w3_2d, det_w3_2d)):
            count2 += 1
    print("Missclassified Percentage in W2 is: " + str(count2*100/len(w1x2)))
    count3 = 0
    for i in range(len(w1x1)):
        a = np.array([w3x1[i], w3x2[i]])
        if q2_g3(a, w3_m_2d, inverse_w3_2d, det_w3_2d) < max(q2_g1(a, w1_m_2d, inverse_w1_2d, det_w1_2d),q2_g3(a, w3_m_2d, inverse_w3_2d, det_w3_2d)):
            count3 += 1
    print("Missclassified Percentage in W3 is: " + str(count3*100/len(w1x3)))
    print("Missclassified Percentage in the dataset: " + str((count1+count2+count3)*100/(len(w1x1)+len(w1x2)+len(w1x3))))
missclassified_qc()

Missclassified Percentage in W1 is: 50.0
Missclassified Percentage in W2 is: 40.0
Missclassified Percentage in W3 is: 100.0
Missclassified Percentage in the dataset: 63.333333333333336


# 3 Features

In [19]:
stacked_w1 = np.stack((w1x1, w1x2, w1x3), axis = 0 )
cov_w1 = np.cov(stacked_w1)
det_w1 = np.linalg.det(cov_w1)
inverse_w1 = np.linalg.inv(cov_w1)

stacked_w2 = np.stack((w2x1, w2x2, w2x3), axis = 0 )
cov_w2 = np.cov(stacked_w2)
det_w2 = np.linalg.det(cov_w2)
inverse_w2 = np.linalg.inv(cov_w2)

stacked_w3 = np.stack((w3x1, w3x2, w3x3), axis = 0 )
cov_w3 = np.cov(stacked_w3)
det_w3 = np.linalg.det(cov_w3)
inverse_w3 = np.linalg.inv(cov_w3)

In [20]:
x = np.array([-5.01, -8.12, -3.68])
print(np.squeeze(q2_g1(x, w1_m, inverse_w1, det_w1)))
print(np.squeeze(q2_g2(x, w2_m, inverse_w2, det_w2)))
print(np.squeeze(q2_g3(x, w3_m, inverse_w3, det_w3)))

-8.896633809823308
-9.88817975608594
-inf


In [21]:
def missclassified_qd():
    count1 = 0
    for i in range(len(w1x1)):
        a = np.array([w1x1[i], w1x2[i], w1x3[i]])
        if q2_g1(a, w1_m, inverse_w1, det_w1) < max(q2_g2(a, w2_m, inverse_w2, det_w2),q2_g3(a, w3_m, inverse_w3, det_w3)):
            count1 += 1
    print("Missclassified Percentage in W1 is: " + str(count1*100/len(w1x1)))
    count2 = 0
    for i in range(len(w1x1)):
        a = np.array([w2x1[i], w2x2[i], w2x3[i]])
        if q2_g2(a, w2_m, inverse_w2, det_w2) < max(q2_g1(a, w1_m, inverse_w1, det_w1),q2_g3(a, w3_m, inverse_w3, det_w3)):
            count2 += 1
    print("Missclassified Percentage in W2 is: " + str(count2*100/len(w1x2)))
    count3 = 0
    for i in range(len(w1x1)):
        a = np.array([w3x1[i], w3x2[i], w3x3[i]])
        if q2_g3(a, w3_m, inverse_w3, det_w3) < max(q2_g1(a, w1_m, inverse_w1, det_w1),q2_g3(a, w3_m, inverse_w3, det_w3)):
            count3 += 1
    print("Missclassified Percentage in W3 is: " + str(count3*100/len(w1x3)))
    print("Missclassified Percentage in the dataset: " + str((count1+count2+count3)*100/(len(w1x1)+len(w1x2)+len(w1x3))))
missclassified_qd()

Missclassified Percentage in W1 is: 20.0
Missclassified Percentage in W2 is: 10.0
Missclassified Percentage in W3 is: 100.0
Missclassified Percentage in the dataset: 43.333333333333336


# Q 2E

For 1 feature : Misclassified Percentage = 53.33% <br>
For 2 features : Misclassified Percentage = 63.33% <br>
For 3 features : Misclassified Percentage = 43.33% <br><br>
The least amount of misclassification happened when all the 3 features were considered.

# Q 2F

In [22]:
def print_class1(x,a) :
    
    v1 = np.squeeze(q2a_g1(x))
    v2 = np.squeeze(q2a_g2(x))
    v3 = np.squeeze(q2a_g3(x))
    print("Value of discriminant function g1 = ", v1)
    print("Value of discriminant function g2 = ", v2)
    print("Value of discriminant function g3 = ", v3)
    if(v1 >= v2 and v1 >= v3):
        print(a, " belongs to class w1")
    elif(v2 >= v1 and v2 >= v3):
        print(a, " belongs to class w2")
    else:
        print(a, " belongs to class w3")

In [23]:
def print_class2(x,a) :
    
    v1 = np.squeeze(q2_g1(x, w1_m_2d, inverse_w1_2d, det_w1_2d))
    v2 = np.squeeze(q2_g2(x, w2_m_2d, inverse_w2_2d, det_w2_2d))
    v3 = np.squeeze(q2_g3(x, w3_m_2d, inverse_w3_2d, det_w3_2d))
    print("Value of discriminant function g1 = ", v1)
    print("Value of discriminant function g2 = ", v2)
    print("Value of discriminant function g3 = ", v3)
    if(v1 >= v2 and v1 >= v3):
        print(a, " belongs to class w1")
    elif(v2 >= v1 and v2 >= v3):
        print(a, " belongs to class w2")
    else:
        print(a, " belongs to class w3")

In [24]:
def print_class3(x) :
    
    v1 = np.squeeze(q2_g1(x, w1_m, inverse_w1, det_w1))
    v2 = np.squeeze(q2_g2(x, w2_m, inverse_w2, det_w2))
    v3 = np.squeeze(q2_g3(x, w3_m, inverse_w3, det_w3))
    print("Value of discriminant function g1 = ", v1)
    print("Value of discriminant function g2 = ", v2)
    print("Value of discriminant function g3 = ", v3)
    if(v1 >= v2 and v1 >= v3):
        print(x, " belongs to class w1")
    elif(v2 >= v1 and v2 >= v3):
        print(x, " belongs to class w2")
    else:
        print(x, " belongs to class w3")
    # print("\n")
    

In [25]:
a = np.array([1,2,1])
b = np.array([5,3,2])
c = np.array([0,0,0])
d = np.array([1,0,0])


In [26]:
print_class1(a[0], a)
print_class1(b[0], b)
print_class1(c[0], c)
print_class1(d[0], d)

Value of discriminant function g1 =  -3.017120237695602
Value of discriminant function g2 =  -3.4475557697828276
Value of discriminant function g3 =  -inf
[1 2 1]  belongs to class w1
Value of discriminant function g1 =  -3.973970790063545
Value of discriminant function g2 =  -3.8323577276183105
Value of discriminant function g3 =  -inf
[5 3 2]  belongs to class w2
Value of discriminant function g1 =  -2.9517539935658155
Value of discriminant function g2 =  -3.4192359531004675
Value of discriminant function g3 =  -inf
[0 0 0]  belongs to class w1
Value of discriminant function g1 =  -3.017120237695602
Value of discriminant function g2 =  -3.4475557697828276
Value of discriminant function g3 =  -inf
[1 0 0]  belongs to class w1


In [27]:
print_class2(a[0:2],a)
print_class2(b[0:2],b)
print_class2(c[0:2],c)
print_class2(d[0:2],d)

Value of discriminant function g1 =  -5.534212026807645
Value of discriminant function g2 =  -5.802956490990918
Value of discriminant function g3 =  -inf
[1 2 1]  belongs to class w1
Value of discriminant function g1 =  -6.229616887740281
Value of discriminant function g2 =  -6.169275647064746
Value of discriminant function g3 =  -inf
[5 3 2]  belongs to class w2
Value of discriminant function g1 =  -5.155657617180433
Value of discriminant function g2 =  -5.530193310472056
Value of discriminant function g3 =  -inf
[0 0 0]  belongs to class w1
Value of discriminant function g1 =  -5.1575542199924245
Value of discriminant function g2 =  -5.546098366604559
Value of discriminant function g3 =  -inf
[1 0 0]  belongs to class w1


In [28]:
print_class3(a)
print_class3(b)
print_class3(c)
print_class3(d)

Value of discriminant function g1 =  -7.926474384871603
Value of discriminant function g2 =  -7.888496475068971
Value of discriminant function g3 =  -inf
[1 2 1]  belongs to class w2
Value of discriminant function g1 =  -8.62373140800979
Value of discriminant function g2 =  -9.061577888334702
Value of discriminant function g3 =  -inf
[5 3 2]  belongs to class w1
Value of discriminant function g1 =  -7.531422860034783
Value of discriminant function g2 =  -7.556398527758285
Value of discriminant function g3 =  -inf
[0 0 0]  belongs to class w1
Value of discriminant function g1 =  -7.530091533434296
Value of discriminant function g2 =  -7.622447688972072
Value of discriminant function g3 =  -inf
[1 0 0]  belongs to class w1


## Inference

When only one and two features were considered, the points (1,2,1) and (5,3,2) were classified as w1 and w2 respectively. But when all the features were considered we found that they were classified as w2 and w1.